### ISSUE TITLE
lint: add enumValue checker

### ISSUE THREAD

**Issue by Quasilyte** [go-critic/go-critic#215]
Detects when literal value is coerced to enum-like type that contains a name for used value.
Suggests using that named constant instead of raw literal.

**Before:**
```go
type color int
const (
  colorDefault color = iota
  colorRed
)
func f() color { return 0 }
```
**After:**
```go
type color int
const (
  colorDefault color = iota
  colorRed
)
func f() color { return colorDefault }
```

**Comment by Quasilyte** [comment]
I've seen multiple such cases in Go sources.
It's not a deal breaker, but does make code harder to understand when there is a big context involved and lots of different constants.
Things like `if cond != 2` is especially awful (suppose cond has some enum-like type and constant for "2" is defined). `if cond != condRepeat` would be much better.

This happens when there were no constants initially and then someone added them.
Some code continues to use raw literals after that, and we get maintenance help.

**Comment by Quasilyte** [comment]
The harder part is determining enum-like types and collecting available named constants set.
Maybe `types.Info` already does this for us.

**Comment by cristaloleg** [comment]
:hugefan:
We should also warn users about comparing enums(> and <) imo.

**Comment by Quasilyte** [comment]
Yep, it's possible and desirable. 

We can check all context that cause untyped value (raw literal) to be converted to typed (named) constant. Only thing we need is a proper map of defined constants.

```go
type enumValue struct {
  Name string // Constant name
  Value constant.Value // from "go/constant"
}
// If we're going to use types.Type as a key, "typeutil.Map" can be useful
// Key is a enum-like type name, value is []enumValue.
// This way, every context that converts ast.BasicLit to a type defined in that map,
// we can check value list: whether it includes used value.
// If list contains definition for that value, use enumValue.Name to make suggestion.
```
https://godoc.org/golang.org/x/tools/go/types/typeutil#Map

**Comment by Quasilyte** [comment]
We have an option of finding all usages of "raw" values of typed constants that have named constants defined for them.
```go
const typed10 myType = 10
/// could use typed10 instead of 10
func f() typedConst { return 10 }
```
